samcli/commands/_utils/click_mutex.py

```diff
@@ -0,0 +1,62 @@
+"""
+Module to check mutually exclusive cli parameters
+"""
+from typing import Any, List, Dict, Tuple
+
+import click
+
+
+class Mutex(click.Option):
+    """
+    Preprocessing checks for mutually exclusive or required parameters as supported by click api.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.required_param_lists: List = kwargs.pop("required_param_lists", None)
+        self.required_params_hint: str = kwargs.pop("required_params_hint", None)
+        self.incompatible_params: List = kwargs.pop("incompatible_params", None)
+        self.incompatible_params_hint: str = kwargs.pop("incompatible_params_hint", None)
+
+        super().__init__(*args, **kwargs)
+
+    def handle_parse_result(self, ctx: click.Context, opts: Dict[str, Any], args: List[str]) -> Tuple[Any, List[str]]:
+        if self.name not in opts:
+            return super().handle_parse_result(ctx, opts, args)
+
+        # Check for parameters not compatible with each other
+        for incompatible_param in self.incompatible_params or []:
+            if incompatible_param in opts:
+                msg = (
+                    f"You must not provide both the {Mutex._to_param_name(self.name)} and "
+                    f"{Mutex._to_param_name(incompatible_param)} parameters.\n"
+                )
+                if self.incompatible_params_hint:
+                    msg += self.incompatible_params_hint
+                raise click.UsageError(msg)
+
+        # Check for required parameters
+        if self.required_param_lists:
+            for required_params in self.required_param_lists:
+                has_all_required_params = False not in [required_param in opts for required_param in required_params]
+                if has_all_required_params:
+                    break
```

Comment:
I couldn't understand the `break` statement here. We are looping for each `required_params` and if we find that one we are breaking the loop, wouldn't this miss the next required param?

Reply:
self.required_param_lists is a list of combinations.
```
    required_param_lists: List[List[str]]
        List of lists with each supported combination of params
        Ex:
        With option = "a" and required_param_lists = [["b", "c"], ["c", "d"]]
        It is valid to specify --a --b --c or --a --c --d
        but not --a --b --d
```

If one combination is satisfied, then we will `break` and skip over `else` block.

Added documentation for the class and this block.

Reply:
I see, so if either `["a", "b", "c"]` or `["a", "b", "d"]` is required, having `["a", "b", "c", "d"]` is still fine because it confirms one or the other requirement. This is not checking it should only be the first case or the second one, right?

Reply:
Yep, that's correct. This is following the previous behavior of `samcli\commands\local\cli_common\click_mutex.py` so that we don't have regressions.